jax/interpreters/pxla.py

```diff
@@ -61,144 +61,138 @@
+            for a in self.mesh_mapping]
+    return (np.broadcast_to(chunk_indices, replica_sizes + chunk_indices.shape)
+              .transpose(perm))
 
 def spec_to_indices(shape: Tuple[int, ...],
-                    sharding_spec: ShardingSpec) -> Tuple[Index, ...]:
-  """Returns numpy-style indices corresponding to sharding_spec.
+                    spec: ShardingSpec) -> Tuple[Index, ...]:
+  """Returns numpy-style indices corresponding to a sharding spec.
 
   Each index describes a shard of the array. The order of the indices is the
   same as the device_buffers of a ShardedDeviceArray (i.e. the data is laid out
-  row-major, with replication treated as an extra innermost dimension).
+  row-major).
 
   Args:
     shape: The shape of the logical array being sharded.
-    sharding_spec: Describes how the array is sharded.
+    spec: Describes how the array is sharded and how the shards are assigned to
+          the logical mesh.
 
   Returns:
-    A tuple of length `prod(sharding_spec.shards_per_axis) *
-    prod(factor for factor, index in sharding_spec.replication_factors)`. Each
-    element is an int, a slice object with step=1, or a tuple thereof, to be
-    treated as an index into the full logical array.
+    A tuple of length equal to the size of the mesh (inferred as the product of
+    sharded dimension sizes and all replication factors).  Each element is an
+    int, a slice object with step=1, or a tuple thereof, to be treated as an
+    index into the full logical array.
   """
-  assert len(shape) == len(sharding_spec.shards_per_axis)
-  if not shape:
-    # special case: scalars can only be indexed by `()`
-    total_replication_factor = int(prod(factor for factor, index in
-                                        sharding_spec.replication_factors))
-    return ((),) * total_replication_factor
-
-  replication_factors = sorted(sharding_spec.replication_factors,
-                               key=op.itemgetter(1))
-  logical_index = 0
-  indices_per_mesh_axis = []
-  for mesh_index in range(len(shape) + len(sharding_spec.replication_factors)):
-    if replication_factors and replication_factors[0][1] == logical_index:
-      # Insert a placeholder `None` to represent a replication factor. These
-      # will all be removed later, since they don't correspond to logical axes.
-      factor, _ = replication_factors.pop(0)
-      indices_per_mesh_axis.append([None] * factor)
-    else:
-      indices = _axis_indices(
-          shape[logical_index],
-          sharding_spec.shards_per_axis[logical_index],
-          sharding_spec.is_axis_materialized[logical_index])
-      indices_per_mesh_axis.append(indices)
-      logical_index += 1
-  assert logical_index == len(shape) and not replication_factors
-
-  indices = list(it.product(*indices_per_mesh_axis))
-
-  # remove placeholder `None`s and trailing colons, then unwrap
-  # single-element tuples
```

Comment:
Yeah it would work with the previous logic, but I still don't understand why do we have to do that. The comment makes it seem like it's a purely cosmetic thing, so why bother?

Reply:
Ohhh is it because we check whether the `__getitem__` argument is contained within indices? Sure I can add that then. Alternatively we could just define a custom equality check for that lookup. We could even optimize it to use a set/map instead of using `in` that will check everything sequentially.

Reply:
Another simple fix would to be always pad the index passed in to `__getitem__` with `slice(None)` to match the rank of the sharded device array. This avoids the branchy canonicalization with a super simple patch there. And if we really care about perf of this indexing then there's already a lot of low hanging fruit there

Reply:
Ok, I moved the (inverse) canonicalization to `__getitem__`, and made sure that indexing into the outermost dimension of `ShardedDeviceArray`s does no copies. I've also added a test to ensure that it's checked by something more than just a comment. See the "Reoptimize ShardedDeviceArray.__getitem__" commit.